### PR TITLE
(proof-new) Finishing work on circuit-propagator

### PR DESCRIPTION
**Pull request by nafur**
This PR finishes the work on the circuit propagator for now. It adds generation of proofs for conflicts and does some general cleanup and refactoring.

### CODE REVIEW

**Comment by nafur** [comment]
@ajreynol The proof that is generated is not yet used. Could you have a look at this?

File: src/theory/booleans/proof_circuit_propagator.cpp [hunk 1]
```diff
@@ -0,0 +1,545 @@
...
+    return mkProof(PfRule::CONTRA, {a, b});
+  }
+  Assert(a->getResult() == b->getResult().notNode());
+  return mkProof(PfRule::CONTRA, {b, a});
+}
+
+std::shared_ptr<ProofNode> ProofCircuitPropagator::andFalse(
+    Node parent, TNode::iterator holdout)
+{
+  if (disabled()) return nullptr;
```
Comment (ajreynol) [inline src/theory/booleans/proof_circuit_propagator.cpp:82]: Does clang format allow this?
I prefer not having if without braces `{ ... }`.

File: src/theory/booleans/proof_circuit_propagator.cpp [hunk 2]
```diff
@@ -0,0 +1,545 @@
...
+std::shared_ptr<ProofNode> ProofCircuitPropagator::eqXFromY(bool y, Node parent)
+{
+  if (disabled()) return nullptr;
+  if (y)
+  {
+    return mkProof(
+        PfRule::EQ_RESOLVE,
+        {assume(parent[1]), mkProof(PfRule::SYMM, {assume(parent)})});
+  }
+  else
```
Comment (ajreynol) [inline src/theory/booleans/proof_circuit_propagator.cpp:125]: This else is unecessary.

File: src/theory/booleans/proof_circuit_propagator.cpp [hunk 3]
```diff
@@ -0,0 +1,545 @@
...
+}
+
+std::shared_ptr<ProofNode> ProofCircuitPropagator::eqYFromX(bool x, Node parent)
+{
+  if (disabled()) return nullptr;
+  if (x)
+  {
+    return mkProof(PfRule::EQ_RESOLVE, {assume(parent[0]), assume(parent)});
+  }
+  else
```
Comment (ajreynol) [inline src/theory/booleans/proof_circuit_propagator.cpp:139]: Ditto.

File: src/theory/booleans/proof_circuit_propagator.cpp [hunk 4]
```diff
@@ -0,0 +1,545 @@
...
+{
+  if (disabled()) return nullptr;
+  if (y)
+  {
+    return mkResolution(
+        mkProof(PfRule::NOT_EQUIV_ELIM2, {assume(parent.notNode())}),
+        {parent[1]},
+        {false});
+  }
+  else
```
Comment (ajreynol) [inline src/theory/booleans/proof_circuit_propagator.cpp:157]: Ditto, more in the rest of the file.

**Review by ajreynol — CHANGES_REQUESTED** [review]
Some minor style issues.

File: src/theory/booleans/proof_circuit_propagator.cpp (hunk 1 above)
Comment (HanielB) [reply src/theory/booleans/proof_circuit_propagator.cpp:82]: I always thought this was a loophole! Clang-format allows it but I think our guidelines are supposed to not allow it?

File: src/theory/booleans/proof_circuit_propagator.cpp [hunk 5]
```diff
@@ -0,0 +1,545 @@
...
+std::shared_ptr<ProofNode> ProofCircuitPropagator::mkResolution(
+    std::shared_ptr<ProofNode> clause,
+    const std::vector<Node>& lits,
+    const std::vector<bool>& polarity)
+{
+  auto* nm = NodeManager::currentNM();
+  std::vector<std::shared_ptr<ProofNode>> children = {clause};
+  std::vector<Node> args;
+  Assert(lits.size() == polarity.size());
+  for (std::size_t i = 0; i < lits.size(); ++i)
```
Comment (HanielB) [inline src/theory/booleans/proof_circuit_propagator.cpp:268]: cache size

File: src/theory/booleans/proof_circuit_propagator.cpp [hunk 6]
```diff
@@ -0,0 +1,545 @@
...
+      else
+      {
+        children.emplace_back(assume(lits[i].notNode()));
+      }
+    }
+    else
+    {
+      children.emplace_back(assume(lits[i]));
+    }
+    args.emplace_back(nm->mkConst<bool>(pol));
```
Comment (HanielB) [inline src/theory/booleans/proof_circuit_propagator.cpp:289]:
```suggestion
    args.emplace_back(nm->mkConst(pol));
```

File: src/theory/booleans/proof_circuit_propagator.cpp (hunk 1 above)
Comment (nafur) [reply src/theory/booleans/proof_circuit_propagator.cpp]: Ha, I hoped you would not complain :-)

**Pull request merged**